src/main/java/com/google/firebase/database/utilities/Utilities.java

```diff
@@ -18,78 +18,117 @@
+        repoInfo.secure = scheme.equals("https") || scheme.equals("wss");
         repoInfo.host += ":" + port;
       } else {
         repoInfo.secure = true;
       }
-      String[] parts = repoInfo.host.split("\\.");
 
-      repoInfo.namespace = parts[0].toLowerCase();
+      Map<String, String> params = getQueryParamsMap(uri.getQuery());
+      String namespaceParam = params.get("ns");
+      if (!Strings.isNullOrEmpty(namespaceParam)) {
+        repoInfo.namespace = namespaceParam;
+      } else {
+        String[] parts = host.split("\\.", -1);
+        repoInfo.namespace = parts[0].toLowerCase();
+      }
+
       repoInfo.internalHost = repoInfo.host;
+
+      String originalPathString = extractPathString(url);
+      // URLEncoding a space turns it into a '+', which is different
+      // from our expected behavior. Do a manual replace to fix it.
+      originalPathString = originalPathString.replace("+", " ");
+      Validation.validateRootPathString(originalPathString);
+
       ParsedUrl parsedUrl = new ParsedUrl();
-      parsedUrl.path = new Path(pathString);
+      parsedUrl.path = new Path(originalPathString);
       parsedUrl.repoInfo = repoInfo;
+
       return parsedUrl;
+    } catch (Exception e) {
+      throw new DatabaseException("Invalid Firebase Database url specified: " + url, e);
+    }
+  }
 
-    } catch (URISyntaxException e) {
-      throw new DatabaseException("Invalid Firebase Database url specified", e);
-    } catch (UnsupportedEncodingException e) {
-      throw new DatabaseException("Failed to URLEncode the path", e);
+  /**
+   * Extracts the path string from the original URL without changing the encoding (unlike
+   * Uri.getPath()).
+   */
+  private static String extractPathString(String originalUrl) {
+    int schemeOffset = originalUrl.indexOf("//");
+    if (schemeOffset == -1) {
+      throw new DatabaseException("Firebase Database URL is missing URL scheme");
+    }
+
+    String urlWithoutScheme = originalUrl.substring(schemeOffset + 2);
+    int pathOffset = urlWithoutScheme.indexOf("/");
+    if (pathOffset != -1) {
+      int queryOffset = urlWithoutScheme.indexOf("?");
+      if (queryOffset != -1) {
+        return urlWithoutScheme.substring(pathOffset + 1, queryOffset);
+      } else {
+        return urlWithoutScheme.substring(pathOffset + 1);
+      }
+    } else {
+      return "";
+    }
+  }
+
+  static Map<String, String> getQueryParamsMap(String queryString) {
+    Map<String, String> paramsMap = new HashMap<>();
+    if (Strings.isNullOrEmpty(queryString)) {
+      return paramsMap;
+    }
+    String[] paramPairs = queryString.split("&");
+    for (String paramPair : paramPairs) {
+      String[] pairParts = paramPair.split("=");
+      String value = paramsMap.get(pairParts[0]);
+      if (Strings.isNullOrEmpty(value)) {
+        value = pairParts[1];
+      } else {
+        value += "," + pairParts[1];
+      }
+      paramsMap.put(pairParts[0], value);
```

Comment:
Do we need to handle any escaping here, or does URI.getQuery() handle this for us?

Reply:
I'm pretty sure we need to unescape in case the URL contains %-encoded characters. `ns` is URL safe so it doesn't break for our use case, but I still want it to be correct as a general utility.

Reply:
Good point @yuchenshi. But it turns out that URI.getQuery() does handle this for us, as demonstrated in a test case that I've added to `TestUtils.parseUrlValid`

So I've added `@VisibleForTesting` to this method, which should restrict its scope of usage as only a companion to `Utilities.parseUrl`